### DISCUSSION TITLE
REST API to OPC UA Server

### DISCUSSION THREAD

**Discussion by shnabin** [FreeOpcUa/opcua-asyncio#582] [Q&A]
I'm new to OPC UA and i have a requirement to read data from a REST API (data in json) & make the data available in OPC UA server, so OPC clients can connect to the server & gets the data. Its more like

REST API --> OPC UA Server --> OPC Clients.

I explored some options to check the possibility & its possible with https://www.opc-router.com/4-rest-protocol-web-services_en/ , I am looking for a similar solution in FreeOpcUa (with out UI etc but as a solution), Can you please suggest how to achieve this.

**Comment by swamper123** [comment]
Hey ho,

at first, remember that this repo is in maintainance mode. You may want to switch over to the most actual one [here](https://github.com/FreeOpcUa/opcua-asyncio).

For the Interface between the REST API and the OPC UA Server, you are on your own. We can only provide help for the server side itself. It sounds that you have to programm something which parses your json data and writes the values into the right nodes on server side with the `set_value()`method of the node (perhaps you may even have to create nodes?). 

Clients can get access on the OPC UA server via TCP-binary (since webservices are neither here nor on the new one realised).

**Reply by shnabin** [reply]
Hi, 

Ok, i will check the recent library.  
I am using python requests library and got the json data, I have almost 2000+ tags in the API, can you direct me to an example to parse multiple values/ create nodes to UA server.

Thanks!!

**Reply by swamper123** [reply]
> Ok, i will check the recent library.

I can transfer the discussion to the new repo if you like to, because things are a bit different there. ;)

Anyway:
I have never worked with *requests* so far (but seems to be interessting, I will take a closer look on that) and I don't know how your data looks like and what it should be in the OPC UA world (a Variable node from Type Int32, Int64, bool, String ...?? Something else?) and I neither know in which nodes you want to store the data (or even create new nodes to store them).
So you are on your own how you interprete/manage/parse your data. Otherwise you have to pay somebody to do that for you. ;)

For the server side stuff: I recommened to have a closer look in the example folder, especially on [server-example.py](https://github.com/FreeOpcUa/python-opcua/blob/master/examples/server-example.py).

With this method from class Node:
https://github.com/FreeOpcUa/python-opcua/blob/3012acf9047aad3acc2674e9008f0d1eee680c25/opcua/common/node.py#L663

You can create a new child (here a variable node) to a parent node (example, the Objects node).
An example is this one from server-example.py
```python
    myarrayvar = myobj.add_variable(idx, "myStronglytTypedVariable", ua.Variant([], ua.VariantType.UInt32))
```

I would recommened this Variant solution, instead of a python value, since the type is otherwise being "guessed". In that example a variable node with an empty array of UInt32 was created.

**Reply by shnabin** [reply]
Hi,

> I can transfer the discussion to the new repo if you like to, because things are a bit different there. ;)

Yes, you can transfer it to new repo. 

> I don't know how your data looks like and what it should be in the OPC UA world (a Variable node from Type Int32, Int64, bool, String ...?? Something else?) 

Multiple data types (APi has 2000 tags) string, float, int32. 
The json data structure is ["MC","EM","Voltage","122.98"] , ["MC","EM","PF","1"] , 
My OPC UA tag should look like MC_EM_Voltage (opc tag) -> 122.98 (value); MC_EM_PF (opc tag) -> 1 (value); etc.

> Otherwise you have to pay somebody to do that for you. ;)

This is always a fallback option :)

I will try your suggestions & try to figure out something. Thank you !!!
